### ISSUE TITLE
Bad date format under certain condition

### ISSUE THREAD

**Issue by ovcharenko** [tobymao/sqlglot#368]
Some weird conversion when format string contains single character, like

```python
>>> sql
'SELECT TO_DATE(boo, "d-M-yyyy HH:mm:ss")'
>>> sqlglot.parse(sql, read='spark')
[(SELECT expressions:
  (TSORDSTODATE this:
    (COLUMN this:
      (IDENTIFIER this: boo, quoted: False)), format:
    (LITERAL this: %-d-%-m-%Y %H:%M:%S, is_string: True)))]
```

**Comment by ovcharenko** [comment]
Here is a [list](https://spark.apache.org/docs/latest/sql-ref-datetime-pattern.html) of Spark supported formats for a reference.

**Comment by tobymao** [comment]
is this fixed in main?

**Comment by samjsharpe** [comment]
No sadly not.

```[work:/Users/sam]$ pip install git+https://github.com/tobymao/sqlglot.git@main
Collecting git+https://github.com/tobymao/sqlglot.git@main
  Cloning https://github.com/tobymao/sqlglot.git (to revision main) to /private/var/folders/cl/6_8xxq9d0md__x881k6c1yx00000gn/T/pip-req-build-654az9xe
  Running command git clone --filter=blob:none --quiet https://github.com/tobymao/sqlglot.git /private/var/folders/cl/6_8xxq9d0md__x881k6c1yx00000gn/T/pip-req-build-654az9xe
  Resolved https://github.com/tobymao/sqlglot.git to commit 84eb3d751f9d76a5dae4654f2ec70bfeae19f957
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: sqlglot
  Building wheel for sqlglot (setup.py) ... done
  Created wheel for sqlglot: filename=sqlglot-5.1.1-py3-none-any.whl size=104543 sha256=be4bdd00f58e216ff4fc771f3954bc617a6a9e3b70729c82116feb4497f3c82d
  Stored in directory: /private/var/folders/cl/6_8xxq9d0md__x881k6c1yx00000gn/T/pip-ephem-wheel-cache-eyb4ivrj/wheels/a5/14/59/1b34dcd580fa00184485256c2bfe2fdf44e18807cf6a2896a0
Successfully built sqlglot
Installing collected packages: sqlglot
Successfully installed sqlglot-5.1.1
WARNING: You are using pip version 22.0.4; however, version 22.2.2 is available.
You should consider upgrading via the '/usr/local/opt/python@3.9/bin/python3.9 -m pip install --upgrade pip' command.
[work:/Users/sam]$ python
>>> import sqlglot
>>> sql = 'SELECT TO_DATE(boo, "d-M-yyyy HH:mm:ss")'
>>> sqlglot.transpile(sql, read='spark', write='spark')
["SELECT TO_DATE(boo, '%-d-%-m-%Y %H:%M:%S')"]
```

I think if I'm reading your excellent code right, then the above comment (`%-d-%-m-%Y %H:%M:%S`) is the *correct* internal format for a time - it looks like you are using this as a canonical representation and that's what the Dialect parser converts the format string to and this is supposed to happen.

What seems to not be happening is the Dialect Generator is not then converting that canonical format back into the dialect-specific format string. I'm guessing this affects Hive and Spark is getting it from inheritance. 

If you can point us in the right direction of where to look I'm happy to do a PR or submit a test, but I'm struggling at the moment to follow what's supposed to be happening (like I figured out that `inverse_time_format` should be involved here and I can see that Dialect.generator deliberately sets `time_mapping: "self.inverse_time_mapping"` which seems like it would be making it the inverse of the parser, so I can't see why that doesn't work.

**Comment by tobymao** [comment]
fixes in v5.1.2

**Comment by samjsharpe** [comment]
Oh wow, thank you very much @tobymao !
